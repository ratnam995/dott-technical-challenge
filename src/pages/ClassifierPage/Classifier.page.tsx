import { MouseEvent, RefObject, useEffect, useRef, useState } from 'react';
import { FileReader } from '../../components/FileReader';
import Gallery from '../../components/Gallery';
import { Overlay } from '../../components/Overlay';
import iDogsAPIResponse from '../../shared/models/DogsAPIResponse.model';
import iPredictions from '../../shared/models/Predictions.model';
import Dogs from '../../shared/services/apis/dogs.api';
import './classifier.page.css';

function ClassifierPage() {
  const [selectedFile, setSelectedFile] = useState<File>();
  const [loadingResult, setLoadingResult] = useState<boolean>(false);
  const [fullList, updateFullList] = useState<Array<string>>([]);
  const [viewGallery, setViewGallery] = useState<boolean>(false);
  const [previewSrc, setPreviewSrc] = useState<string>();

  const imgFormRef: RefObject<HTMLDivElement> = useRef() as RefObject<HTMLDivElement>;
  const imgGalleryRef: RefObject<HTMLDivElement> = useRef() as RefObject<HTMLDivElement>;
  
  useEffect(() => {
    if (!selectedFile) {
        setPreviewSrc(undefined)
        return
    }

    const objectUrl: string = URL.createObjectURL(selectedFile)
    setPreviewSrc(objectUrl)

    // free memory when ever this component is unmounted
    return () => URL.revokeObjectURL(objectUrl)
  }, [selectedFile])

  useEffect(() => {
    if(viewGallery) {
      if(imgGalleryRef.current) {
        imgGalleryRef.current.scrollIntoView({ 
          behavior: 'smooth' 
        });
      }
    } else {
      if(imgFormRef.current) {
        imgFormRef.current.scrollIntoView({ 
          behavior: 'smooth' 
        });
      }
    }
  }, [viewGallery])

  const fetchFullGallery = async (predictions: Array<iPredictions>) => {
    const [ subBreed, breed ] = predictions[0].className.split(" ");
    Dogs.getAll(breed, subBreed)
    .then((response: iDogsAPIResponse) => {
      if(response) {
        if(response.code === 404) {
          alert("Is it a dinosaur? Doesn't look like a dog. Please upload a different image.")
        } else {
          updateFullList([...response.message]);
        }
      } else {
        alert("Oops! this wasn't expected. Please try again.")
      }
    })
    .catch((e) => {
      alert("Oops! this wasn't expected. Please try again.")
    })
    .finally(() => {
      setLoadingResult(false);
    });
  }

  const toggleViewGallery = (e: MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    setViewGallery(!viewGallery);
  }

  return (
    <div className="scroll-container">
      {
        loadingResult ? <Overlay/>
        : null
      }
      {
        viewGallery ? null
        : <FileReader
          ref={imgFormRef}
          selectedFile={selectedFile}
          previewSrc={previewSrc}
          setSelectedFile={setSelectedFile}
          fetchFullGallery={fetchFullGallery}
          setLoadingResult={setLoadingResult}
        />
      }
      {
        !loadingResult && fullList && fullList.length ?
          <button
            className={`view-gallery-cta ${viewGallery ? 'top' : 'bottom'}`}
            onClick={toggleViewGallery}
            dangerouslySetInnerHTML={{ __html: viewGallery ? "View form &#8593;" : "View gallery &#8595;"}}
          />
          : null
      }
      {
        viewGallery && fullList && fullList.length ? <Gallery fullList={fullList} ref={imgGalleryRef}/>
        : null
      }
    </div>
  );
}

export default ClassifierPage;
