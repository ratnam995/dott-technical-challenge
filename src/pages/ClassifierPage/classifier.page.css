.wrapper {
    width: 100%;
    height: 100vh;
    display: flex;
    flex-direction: row;
    align-items: center;
}

.form-wrapper {
    width: 50%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    border-right: 2px dotted gray;
}

#form-file-upload {
    height: 16rem;
    width: 28rem;
    max-width: 100%;
    text-align: center;
    position: relative;
    cursor: pointer;
}

#input-file-upload {
    display: none;
}

#label-file-upload {
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    border-width: 2px;
    border-radius: 1rem;
    border-style: dashed;
    border-color: #cbd5e1;
    background-color: #f8fafc;
}

.upload-button {
    cursor: pointer;
    padding: 0.25rem;
    font-size: 1rem;
    border: none;
    font-family: 'Oswald', sans-serif;
    background-color: transparent;
}

.upload-button:hover {
    text-decoration-line: underline;
}

#label-file-upload.drag-active {
    background-color: #ffffff;
}

.preview-wrapper {
    width: 50%;
    height: 100%;
    padding: 100px 20px 100px 20px;
    box-sizing: border-box;
    display: flex;
    align-items: center;
    justify-content: center;
}

.image-preview {
    max-width: 100%;
    max-height: 100%;
    box-sizing: border-box;
}

.confirm-btn {
    width: 100px;
    height: 40px;
    background-color: green;
    color: white;
    border: none;
    border-radius: 5px;
    margin-top: 20px;
    font-size: 16px;
    cursor: pointer;
}

.confirm-btn:hover {
    transform: scale(105%);
    transition: transform 0.5s;
    transition-timing-function: ease-in-out;
}

@keyframes bounce {
    0% {
        transform: scale(1,1) translate(-50%, 0px);
    }

    30%{
        transform: scale(1,0.8) translate(-50%, 10px); 
    }

    75%{
        transform: scale(1,1.1) translate(-50%, -15px); 
    }

    100% {
        transform: scale(1,1) translate(-50%, 0px);
    }
}

.view-gallery-cta {
    position: fixed;
    left: 50%;
    border: none;
    background-color: transparent;
    text-align: center;
    font-size: 18px;
    font-weight: 600;
    cursor: pointer;
    animation: bounce 1s infinite;
    height: 100px;
}

.view-gallery-cta:hover {
    color: darkmagenta;
}

.top {
    top: 0px;
}

.bottom {
    bottom: 0px;
}

.image-container > img {
    max-width: 300px;
}

@media screen and (max-width: 560px) {
    .wrapper {
        flex-direction: column;
    }

    .form-wrapper {
        width: 100%;
        border-right: unset;
        border-bottom: 2px dotted gray;
        padding: 10px;
    }

    .preview-wrapper {
        width: 100%;
        padding: 50px 10px;
    }
}