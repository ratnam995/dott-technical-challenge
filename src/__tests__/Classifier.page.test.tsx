import renderer from 'react-test-renderer';
import ClassifierPage from '../pages/ClassifierPage/Classifier.page';

import { BrowserRouter } from 'react-router-dom';

describe("Classifier page", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(<BrowserRouter><ClassifierPage/></BrowserRouter>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});