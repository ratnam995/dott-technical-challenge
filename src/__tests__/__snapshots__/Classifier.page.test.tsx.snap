// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Classifier page renders correctly 1`] = `
<div
  className="scroll-container"
>
  <div
    className="form-container wrapper"
  >
    <div
      className="form-wrapper"
    >
      <form
        id="form-file-upload"
        onDragEnter={[Function]}
        onDrop={[Function]}
        onSubmit={[Function]}
      >
        <input
          id="input-file-upload"
          multiple={false}
          onChange={[Function]}
          type="file"
        />
        <label
          className=""
          htmlFor="input-file-upload"
          id="label-file-upload"
        >
          <div>
            <p>
              Drag and drop your file here or
            </p>
            <button
              className="upload-button"
              onClick={[Function]}
            >
              Upload a file
            </button>
          </div>
        </label>
        <button
          className="confirm-btn"
          onClick={[Function]}
        >
          Confirm
        </button>
      </form>
    </div>
    <div
      className="preview-wrapper"
    />
  </div>
</div>
`;
